File: src/main/java/io/playground/orderservice/infrastructure/persistence/saga/ProcessSagaPersistenceAdapter.java
package io.playground.orderservice.infrastructure.persistence.saga;

import io.playground.orderservice.application.saga.port.persistence.ProcessSagaPersistencePort;
import io.playground.orderservice.domain.saga.ProcessSaga;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class ProcessSagaPersistenceAdapter implements ProcessSagaPersistencePort {
    private final ProcessSagaJpaRepository processSagaRepository;
    private final NamedParameterJdbcTemplate jdbcTemplate;

    @Override
    public Optional<ProcessSaga> findByOrderExternalId(String orderExternalId) {
        return processSagaRepository.findByOrderExternalId(orderExternalId)
                .map(ProcessSagaEntity::toDomain);
    }

    @Override
    public ProcessSaga saveAndFlush(ProcessSaga processSaga) {
        return processSagaRepository.saveAndFlush(
                ProcessSagaEntity.fromDomain(processSaga)
        ).toDomain();
    }

    @Override
    public List<ProcessSaga> findExpiredSagas(int limitSize,
                                              int retryMax) {
        return processSagaRepository
                .findExpiredSagas(limitSize, retryMax).stream()
                .map(ProcessSagaEntity::toDomain)
                .toList();
    }

    @Override
    public boolean updateStatus(Long id,
                                ProcessSaga.ProcessSagaStatus status) {
        return ProcessSaga.isActiveStatus(status) ?
                jdbcTemplate.update(
                        "UPDATE process_sagas " +
                                "SET status = :status " +
                                "WHERE id = :id",
                        Map.of(
                                "id", id,
                                "status", status.name()
                        )
                )== 1 :
                jdbcTemplate.update(
                        "UPDATE process_sagas " +
                                "SET status = :status, " +
                                    "is_active = false " +
                                "WHERE id = :id",
                        Map.of(
                                "id", id,
                                "status", status.name()
                        )
                )== 1;
    }

    @Override
    public boolean updateStatus(String orderExternalId,
                                ProcessSaga.ProcessSagaStatus status,
                                ProcessSaga.ProcessSagaStatus beforeStatus) {
        return ProcessSaga.isActiveStatus(status) ?
                jdbcTemplate.update(
                        "UPDATE process_sagas " +
                                "SET status = :status " +
                            "WHERE order_external_id = :orderExternalId " +
                                "AND status = :beforeStatus",
                        Map.of(
                                "orderExternalId", orderExternalId,
                                "status", status.name(),
                                "beforeStatus", beforeStatus.name()
                        )
                ) == 1 :
                jdbcTemplate.update(
                        "UPDATE process_sagas " +
                                "SET status = :status, " +
                                    "is_active = false " +
                            "WHERE order_external_id = :orderExternalId " +
                                "AND status = :beforeStatus",
                        Map.of(
                                "orderExternalId", orderExternalId,
                                "status", status.name(),
                                "beforeStatus", beforeStatus.name()
                        )
                ) == 1;
    }

    @Override
    public boolean updatePaymentKey(String orderExternalId,
                                    String paymentKey) {
        return jdbcTemplate.update(
                "UPDATE process_sagas " +
                        "SET payment_key = :paymentKey " +
                    "WHERE order_external_id = :orderExternalId",
                Map.of(
                        "orderExternalId", orderExternalId,
                        "paymentKey", paymentKey
                )
        ) == 1;
    }

    @Override
    public boolean updateRetryCountAndLockedUntil(Long id, Instant lockedUntil) {
        return jdbcTemplate.update(
                "UPDATE process_sagas " +
                        "SET retry_count = retry_count + 1, " +
                        "AND locked_until = :lockedUntil " +
                    "WHERE id = :id",
                Map.of(
                        "id", id,
                        "locked_until", lockedUntil
                )
        ) == 1;
    }
}

File: src/main/kotlin/io/playground/orderservice/infrastructure/persistence/saga/ProcessSagaPersistenceAdapter.kt
package io.playground.orderservice.infrastructure.persistence.saga

import io.playground.orderservice.application.saga.port.persistence.ProcessSagaPersistencePort
import io.playground.orderservice.domain.saga.ProcessSaga
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.Optional

@Component
class ProcessSagaPersistenceAdapter(
    private val processSagaRepository: ProcessSagaJpaRepository,
    private val jdbcTemplate: NamedParameterJdbcTemplate,
) : ProcessSagaPersistencePort {
    override fun findByOrderExternalId(orderExternalId: String): Optional<ProcessSaga> =
        processSagaRepository.findByOrderExternalId(orderExternalId)
            .map(ProcessSagaEntity::toDomain)

    override fun saveAndFlush(processSaga: ProcessSaga): ProcessSaga =
        processSagaRepository.saveAndFlush(ProcessSagaEntity.fromDomain(processSaga)).toDomain()

    override fun findExpiredSagas(limitSize: Int, retryCount: Int): List<ProcessSaga> =
        processSagaRepository.findExpiredSagas(limitSize, retryCount)
            .map(ProcessSagaEntity::toDomain)

    override fun updateStatus(id: Long?, status: ProcessSaga.ProcessSagaStatus): Boolean =
        if (ProcessSaga.isActiveStatus(status)) {
            jdbcTemplate.update(
                "UPDATE process_sagas SET status = :status WHERE id = :id",
                mapOf(
                    "id" to id,
                    "status" to status.name,
                ),
            ) == 1
        } else {
            jdbcTemplate.update(
                "UPDATE process_sagas SET status = :status, active = false WHERE id = :id",
                mapOf(
                    "id" to id,
                    "status" to status.name,
                ),
            ) == 1
        }

    override fun updateStatus(
        orderExternalId: String,
        status: ProcessSaga.ProcessSagaStatus,
        beforeStatus: ProcessSaga.ProcessSagaStatus,
    ): Boolean = if (ProcessSaga.isActiveStatus(status)) {
        jdbcTemplate.update(
            "UPDATE process_sagas SET status = :status WHERE order_external_id = :orderExternalId AND status = :beforeStatus",
            mapOf(
                "orderExternalId" to orderExternalId,
                "status" to status.name,
                "beforeStatus" to beforeStatus.name,
            ),
        ) == 1
    } else {
        jdbcTemplate.update(
            "UPDATE process_sagas SET status = :status, active = false WHERE order_external_id = :orderExternalId AND status = :beforeStatus",
            mapOf(
                "orderExternalId" to orderExternalId,
                "status" to status.name,
                "beforeStatus" to beforeStatus.name,
            ),
        ) == 1
    }

    override fun updatePaymentKey(orderExternalId: String, paymentKey: String): Boolean =
        jdbcTemplate.update(
            "UPDATE process_sagas SET payment_key = :paymentKey WHERE order_external_id = :orderExternalId",
            mapOf(
                "orderExternalId" to orderExternalId,
                "paymentKey" to paymentKey,
            ),
        ) == 1

    override fun updateRetryCountAndLockedUntil(id: Long?, lockedUntil: Instant?): Boolean =
        if (lockedUntil != null) {
            jdbcTemplate.update(
                "UPDATE process_sagas SET retry_count = retry_count + 1, locked_until = :lockedUntil WHERE id = :id",
                mapOf(
                    "id" to id,
                    "lockedUntil" to lockedUntil,
                ),
            ) == 1
        } else {
            jdbcTemplate.update(
                "UPDATE process_sagas SET retry_count = retry_count + 1, locked_until = null WHERE id = :id",
                mapOf("id" to id),
            ) == 1
        }
}
